### ISSUE TITLE
SQL State table overwhelmed when Semaphores are full

### ISSUE THREAD

**Issue by scargo** [HangfireIO/Hangfire#1858]
The following three rows are put into the Hangfire.State table for each JobId that is waiting on a semaphore as fast as possible.

Id	JobId	Name	Reason	CreatedAt	Data
1800693	281335	Enqueued	Triggered by DelayedJobScheduler	2021-05-12 22:34:41.510	{"EnqueuedAt":"2021-05-12T22:34:41.5102635Z","Queue":"passive"}
1800697	281335	Processing	NULL	2021-05-12 22:34:41.517	{"StartedAt":"2021-05-12T22:34:41.5149262Z","ServerId":"#####","WorkerId":"c02347ca-c92a-4783-97b1-57f91a31e341"}
1800698	281335	Scheduled	Couldn't acquire Semaphore("#####"): limit exceeded	2021-05-12 22:34:41.517	{"EnqueueAt":"2021-05-12T22:35:41.5170123Z","ScheduledAt":"2021-05-12T22:34:41.5170123Z"}

This quickly creates millions of rows in the table.  This table alone quickly became 20GB with 10K jobs waiting to be processed.  We have batches of data that are waiting on a limited number of API resources available.

It seems that this continual logging of information is unnecessary.  The job should probably not change state if there is not a semaphore available, rather than changing state 3 times in quick succession and repeating that pattern as fast as possible.

**Comment by odinserj** [comment]
Please use the following query first to reduce the number of state entries in your [Hangfire].[State] table. This query removes all state entries except the last 10, and removes them in batches of 50,000 records – otherwise expiration manager will try to remove all the state records at once, and this will be inefficient if a particular background job have thousands of them.

```sql
SET TRANSACTION ISOLATION LEVEL READ COMMITTED;
SET XACT_ABORT ON;
SET LOCK_TIMEOUT 1000;
WHILE 1 = 1
BEGIN
  ;WITH GROUPED AS (
    SELECT *, ROW_NUMBER() OVER (PARTITION BY [JobId] ORDER BY [JobId], [Id] DESC) AS RowNo 
    FROM [HangFire].[State]
  )
  DELETE TOP (50000) FROM GROUPED WHERE RowNo > 10
  IF @@ROWCOUNT = 0 BREAK;

  WAITFOR DELAY '00:00:01';
END
```

After removing the entries please tell me what re-scheduling delay you are using for your semaphores? Are you using the default one (1 minute) or have a custom one (like zero seconds)? I'm afraid that temporary workaround would be to increase this delay, while I'm implementing a better solution.

Next temporary solution I will implement will be the same as it's already implemented in Hangfire.Pro.Redis – to limit the number of persisted state entries to a some predefined number, for example, 10 like in the script above (but this will force another query to be performed each time a new state entry is added). Another solution I have is to make it possible to reuse the last state entry for some scenarios, but its applicability can be minimal for some cases, so I need to think more.

**Comment by scargo** [comment]
We can increase the delay, but we'd rather not make it big enough where it will solve the problem.  We have 10 Dockers that run a document processing process that takes quite a while to process each one (1-10 minutes per call) and we want to keep them running as much as possible.  We have a backlog of documents that need to go through them now but when we don't have a backlog we want them to go in as soon as possible.  

Until there is a better fix, we'll probably just schedule this delete job to run periodically to keep the size down.

Thanks.

**Comment by odinserj** [comment]
Have you considered to move these document processing jobs to a dedicated queue with some predefined number of workers?

**Comment by scargo** [comment]
We have.  We like the idea of using a lower priority queue for these things so that when other tasks are flooding the system these tasks can be put on the back burner.  Also we have two versions of these tasks so the ability to use database level semaphore values to adjust the dials is really nice.  We can have normal operations 9 live and 1 backlog and we can switch that to 5/5 just by updating the DB when we have some analytic test we need to update.  We are looking at upgrading the 10 dockers we have to 20 assuming this proof of concept goes well.

These are all workarounds.  I think there is a fundamental problem in Hangfire when the job changes state 3 times when there is no semaphore available to run it.
